crates/e2e/tests/e2e/protocol_fee.rs

```diff
@@ -16,413 +21,426 @@ use {
-    let fee_policy = FeePolicyKind::Surplus {
-        factor: 0.3,
-        max_volume_factor: 0.9,
-    };
-    let protocol_fee = ProtocolFee {
-        policy: fee_policy,
-        policy_order_class: FeePolicyOrderClass::Market,
+    tracing::info!("Waiting for orders metadata update.");
+    let metadata_updated = || async {
+        onchain.mint_block().await;
+        let market_order_updated = services
+            .get_order(&market_price_improvement_uid)
+            .await
+            .is_ok_and(|order| !order.metadata.executed_surplus_fee.is_zero());
+        let limit_order_updated = services
+            .get_order(&limit_surplus_order_uid)
+            .await
+            .is_ok_and(|order| !order.metadata.executed_surplus_fee.is_zero());
+        let partner_fee_order_updated = services
+            .get_order(&partner_fee_order_uid)
+            .await
+            .is_ok_and(|order| !order.metadata.executed_surplus_fee.is_zero());
+        market_order_updated && limit_order_updated && partner_fee_order_updated
     };
-    // Without protocol fee:
-    // Expected execution is 5040413426236634210 GNO for 5000000000000000000 DAI,
-    // with executed_surplus_fee = 167058994203399 GNO
-    //
-    // With protocol fee:
-    // surplus in sell token = 10000000000000000000 - 5040413426236634210 =
-    // 4959586573763365790
-    //
-    // protocol fee in sell token = 0.3*4959586573763365790 = 1487875972129009737
-    //
-    // expected executed_surplus_fee is 167058994203399 + 1487875972129009737 =
-    // 1488043031123213136
-    //
-    // Settlement contract balance after execution = executed_surplus_fee GNO
-    execute_test(
-        web3.clone(),
-        vec![ProtocolFeesConfig(vec![protocol_fee])],
-        OrderKind::Buy,
-        None,
-        1488043031123213136u128.into(),
-        1488043031123213136u128.into(),
-    )
-    .await;
+    wait_for_condition(TIMEOUT, metadata_updated).await.unwrap();
+
+    let market_price_improvement_order = services
+        .get_order(&market_price_improvement_uid)
+        .await
+        .unwrap();
+    let market_executed_surplus_fee_in_buy_token =
+        market_price_improvement_order.metadata.executed_surplus_fee
+            * market_quote_after.buy_amount
+            / market_quote_after.sell_amount;
+    let market_quote_diff = market_quote_after
+        .buy_amount
+        .saturating_sub(market_quote_before.quote.buy_amount);
+    assert!(market_executed_surplus_fee_in_buy_token >= market_quote_diff * 3 / 10);
+
+    let partner_fee_order = services.get_order(&partner_fee_order_uid).await.unwrap();
+    let partner_fee_executed_surplus_fee_in_buy_token =
+        partner_fee_order.metadata.executed_surplus_fee * partner_fee_quote_after.buy_amount
+            / partner_fee_quote_after.sell_amount;
+    assert!(
+        partner_fee_executed_surplus_fee_in_buy_token
+            >= partner_fee_quote_after.buy_amount * 2 / 100
```

Comment:
If someone changes the fee policy above this test may pass because there is no upper bound check. Can we somehow approximate this?

Reply:
The difference here comes from the network fee, which is computed by the baseline solver. Not sure it is possible to predict the value based on the given data. The value can be hardcoded. Wdyt?

Reply:
After reading the test again, I think the fact that we exactly assert on the settlement balances ensures that even if we drastically overcomputed the fee, the test would fail, so I think it's fine.